pykern/fconf.py

```diff
@@ -0,0 +1,569 @@
+            base = []
+        res = []
+        for i, e in enumerate(new):
+            # lists don't have base values
+            with self._xpath(f"[{i}]"):
+                x = self._do(e, None)
+            if isinstance(x, list):
+                res.extend(x)
+            else:
+                res.append(x)
+        return res + base
+
+    @contextlib.contextmanager
+    def _xpath(self, element=None):
+        """Track element path for errors
+
+        If there is an exception, xpath will hold the place of the exception
+        """
+        if element is not None:
+            self.xpath.push(element)
+            pkdc("{}", element)
+            yield
+            self.xpath.pop()
+        else:
+            yield
+
+
+class _File(PKDict):
+    def __str__(self):
+        return "source=" + pykern.pkio.py_path().bestrelpath(self.path)
+
+
+class _Macro(PKDict):
+    def call(self, namespace, args, kwargs):
+        return self.func(namespace, *args, **kwargs)
+
+    def _kind(self):
+        return "pymacro"
+
+    def __str__(self):
+        return f"{self._kind}={self.name} {self.source}"
+
+
+class _Namespace:
+    def __init__(self, evaluator):
+        self._evaluator = evaluator
+
+    def __func(self, name, exc):
+        m = self._evaluator.parser.macros.get(name)
+        if not m:
+            raise exc(f"macro name={name}")
+
+        def x(*args, **kwargs):
+            a = list(args)
+            if a and a[0] is self:
+                a.pop(0)
+            return m.call(self, a, kwargs)
+
+        return x
+
+    def __getattr__(self, name):
+        return self.__func(name, AttributeError)
+
+    def __getitem__(self, name):
+        if name == _SELF:
+            return self
+        return self.__func(name, KeyError)
+
+
+class _XPath:
+    def __init__(self):
+        self.stack = []
+
+    def push(self, key):
+        f = inspect.currentframe().f_back.f_back.f_back
+        self.stack.append(
+            PKDict(key=key, func=f.f_code.co_name, line=f.f_lineno),
+        )
+
+    def pop(self):
+        self.stack.pop()
+
+    def top(self):
+        if self.stack:
+            return self._str(self.stack[-1])
+        else:
+            "None"
+
+    def __str__(self):
+        return "/".join([k.key for k in self.xpath.stack])
+
+    def stack_as_str(self):
+        res = "Evaluator stack:\n"
+        for i in self.stack:
+            res += self._str(i)
+        return res
+
+    def _str(self, element):
+        return f"{element.func}:{element.line} {str(element.key):.100}\n"
+
+
+class _YAMLMacro(PKDict):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # TEST: duplicate params, invalid params, etc.
+        s = set()
+        for p in self.params:
+            if p in s:
+                raise TypeError(f"duplicate param={p} in macro={self.name}")
+            s.add(p)
+
+    def call(self, namespace, args, kwargs):
+        return namespace._evaluator.start(
```

Comment:
`_evaluator` should be public if accessed like this

Reply:
See https://github.com/radiasoft/pykern/pull/224/files#r915385141